lib/libc/audio/lib_buffer.c

```diff
@@ -0,0 +1,208 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <debug.h>
+
+#include <tinyara/audio/audio.h>
+#include "libc.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Configuration ************************************************************/
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: apb_semtake
+ *
+ *       Take an Audio Pipeline Buffer.
+ *
+ ****************************************************************************/
+
+static void apb_semtake(FAR struct ap_buffer_s *apb)
+{
+	/* Take the semaphore (perhaps waiting) */
+
+	while (sem_wait(&apb->sem) != 0) {
+		/* The only case that an error should occr here is if
+		 * the wait was awakened by a signal.
+		 */
+
+		ASSERT(errno == EINTR);
+	}
+}
+
+/****************************************************************************
+ * Name: apb_semgive
+ ****************************************************************************/
+
+#define apb_semgive(b) sem_post(&b->sem)
+
+/****************************************************************************
+ * Name: apb_alloc
+ *
+ * Allocate an Audio Pipeline Buffer for use in the Audio sub-system.  This
+ * will perform the actual allocate based on buffer data format, number of
+ * channels, etc. and prepare the buffer for consumption.
+ *
+ ****************************************************************************/
+
+int apb_alloc(FAR struct audio_buf_desc_s *bufdesc)
+{
+	uint32_t bufsize;
+	int ret;
+	struct ap_buffer_s *apb;
+
+	DEBUGASSERT(bufdesc->u.ppBuffer != NULL);
```

Comment:
It would be better to check whether bufdesc pointer is valid or not before dererencing as it's a public api as follows
DEBUGASSERT(bufdesc != NULL && bufdesc->u.ppBuffer != NULL);